server/src/instant/grouped_queue.clj

```diff
@@ -0,0 +1,53 @@
+(ns instant.grouped-queue
+  (:import
+   (java.util.concurrent LinkedBlockingQueue)
+   (clojure.lang PersistentQueue)))
+
+(def empty-q PersistentQueue/EMPTY)
+
+(defn create [group-fn]
+  {:group-fn group-fn
+   :group-key->subqueue (atom {})
+   :main-queue (LinkedBlockingQueue.)})
+
+(defn enqueue! [{:keys [group-fn main-queue group-key->subqueue]
+                 :as _grouped-q} item]
+  (let [group-key (group-fn item)]
+    (if (nil? group-key)
+      ;; This item is not to be grouped.
+      (.put main-queue [:item item])
+      ;; This item will be grouped on `group-key` 
+      (let [[prev] (swap-vals! group-key->subqueue
+                               update
+                               group-key
+                               (fn [subqueue]
+                                 (conj (or subqueue empty-q)
+                                       item)))
+            prev-subqueue (get prev group-key)
+            first-enqueue? (empty? prev-subqueue)]
+        (when first-enqueue?
+          (.put main-queue [:group-key group-key]))))))
+
+(defn process! [{:keys [main-queue group-key->subqueue]
+                 :as _grouped-q} process-fn]
+  (let [[t arg] (.take main-queue)]
+    (case t
+      :item
+      (let [item arg]
+        (process-fn item))
+      :group-key
+      (let [group-key arg
+            item (first (get @group-key->subqueue group-key))
+            _ (process-fn item)
+            curr (swap! group-key->subqueue update group-key pop)
```

Comment:
Note: `swap!` happens _after_ `process-fn`, to avoid a race condition. 

I originally wrote: 

```clojure
      (let [group-key arg
            [prev curr] (swap-vals! group-key->subqueue update group-key pop)
            prev-subqueue (get prev group-key)
            item (first prev-subqueue)
            curr-subqueue (get curr group-key)]
        (process-fn item)
        (when (seq curr-subqueue)
          (.put main-queue [:group-key group-key]))))))
```

But I realized this could lead to a race condition, where two threads were handling items for the same `group-key`: 

1. Imagine a group size = 1, 2 workers
2. Worker1 runs process!, does swap-vals!, and starts process-fn
3. While that is happening, we receive an enqueue! 
4. Now, enqueue would see that we went from no elements to one element, and add to main-queue
5. Worker2 would pick it up, and voila, 2 items are running in parallel for the same group-key